### ISSUE TITLE
Unnecessary duplicate module installations 

### ISSUE THREAD

**Issue by ZeeCoder** [yarnpkg/yarn#5171]
**bug** 
Might not acutally be a bug, just a different behaviour from NPM?
Regardless it was unexpected for me.

**What is the current behavior?**
I have a small lib that requires postcss@6.0.9 exactly.
A dependency accepts postcss@^6.0.1.
I thought yarn would just install 6.0.9, since that single version satisfies both constraints, but instead 2 versions get installed: 6.0.9 and also the latest one: 6.0.16.

The problem with this approach, is that I have an issue with postcss@6.0.10 and above, so I hoped I could just specify an exact version instead on the root level.

**If the current behavior is a bug, please provide the steps to reproduce.**
Just clone https://github.com/ZeeCoder/cq-demo-utils and run yarn.

**What is the expected behavior?**

I thought it would just install postcss@6.0.9, as specified in root, since that would be acceptable for the dependency too.

**Please mention your node.js, yarn and operating system version.**

yarn: 1.3.2
Node: 8.1.3

**Comment by KidkArolis** [comment]
Yeah, somewhat strange, I thought yarn dedupes by default. Looks like it only does so if you run `yarn --flat`.

You can add this to your `package.json` as a workaround (which is what `yarn --flat` adds as well):

```
"resolutions": {
  "postcss": "6.0.9"
}
```

**Comment by ZeeCoder** [comment]
Interesting, I'll try that.

I'm not sure if codesandbox.io runs yarn or not, because it seems to do the same.

That's the reason I've noticed the issue.
Unfortunately I have no control over the package manager there though. 🤔😣

**Comment by Silic0nS0ldier** [comment]
Just hit with issue with using `flat`.
![image](https://user-images.githubusercontent.com/17376090/34857956-bd2ab5b2-f7a1-11e7-96a1-c7684a04086a.png)

It should have resolved to 2.2.4, since it was the newest release that satisfied all ranges. NPM was able to properly resolve to 2.2.4 with no duplicates.

`package.json` follows:
```json
{
  "private": true,
  "dependencies": {
    "bootstrap": "^3.3.6",
    "clipboard": "^1.5.16",
    "font-awesome": "^4.7.0",
    "handlebars": "^3.0.3",
    "jquery": "^2.2.4",
    "jquery-ui": "^1.12.0",
    "jquery-validation": "~1.14.0",
    "select2": "^4.0.5",
    "tablesorter": "^2.28.5",
    "moment": "^2.17.1",
    "speakingurl": "^11.0.0",
    "icheck": "^1.0.1",
    "jquery-slimscroll": "~1.3.8",
    "fastclick": "^1.0.6",
    "ionicons": "^2.0.1",
    "urijs": "^1.18.4"
  },
  "resolutions": {
    "optimist": "0.6.1"
  }
}

```

**Comment by Silic0nS0ldier** [comment]
Encountered this with another dependency. Project uses `moment` with semver range of `^2.17.1`, with a dependency looking for `~2.18.0`. Should be fine, as 2.18.0 or 2.18.1 satisfy this. Except instead 2.18.1 and 2.20.1 are installed.

**Comment by arcanis** [comment]
Yarn *does* dedupe (even without --flat), but we have some issues in some cases (hoisting is hard, and our code is quite complex). We have medium-term plans that will mitigate that but in the meantime, the resolution override is your best shot (the second best one is to open a PR in the affected project to bump the version so that you don't suffer from the bug).

**Comment by ZeeCoder** [comment]
I worked around it for the moment, so I'm fine waiting for the medium-term plans to bore their fruits :)

**Comment by StephanBijzitter** [comment]
I have another case of deduping failing:
```
"@types/node@*":
  version "9.4.6"
  resolved "https://registry.yarnpkg.com/@types/node/-/node-9.4.6.tgz#d8176d864ee48753d053783e4e463aec86b8d82e"

"@types/node@^8.9.1":
  version "8.9.4"
  resolved "https://registry.yarnpkg.com/@types/node/-/node-8.9.4.tgz#dfd327582a06c114eb6e0441fa3d6fab35edad48"

"@types/react-dom@^16.0.4":
  version "16.0.4"
  resolved "https://registry.yarnpkg.com/@types/react-dom/-/react-dom-16.0.4.tgz#2e8fd45f5443780ed49bf2cdd9809e6091177a7d"
  dependencies:
    "@types/node" "*"
    "@types/react" "*"
```
In this case `@types/node v9.4.6` should never be installed
